File: android/app/src/main/java/com/simplelink/app/Protocol.kt
package com.simplelink.app

import org.json.JSONObject
import java.io.DataOutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder

enum class MessageType(val id: Byte) {
    AUTH(1),
    AUTH_OK(2),
    AUTH_FAIL(3),
    CLIPBOARD(4),
    FILE_BEGIN(5),
    FILE_CHUNK(6),
    FILE_END(7),
    PING(8),
    PONG(9);

    companion object {
        fun from(id: Byte): MessageType? = entries.find { it.id == id }
    }
}

object LinkProtocol {
    val magic = "SLNK".toByteArray()
    const val DEFAULT_PORT = 9473
    const val CHUNK_SIZE = 64 * 1024

    fun encode(type: MessageType, payload: ByteArray = ByteArray(0)): ByteArray {
        val buffer = ByteBuffer.allocate(9 + payload.size).order(ByteOrder.BIG_ENDIAN)
        buffer.put(magic)
        buffer.put(type.id)
        buffer.putInt(payload.size)
        buffer.put(payload)
        return buffer.array()
    }

    fun jsonObject(payload: ByteArray): JSONObject? =
        runCatching { JSONObject(String(payload, Charsets.UTF_8)) }.getOrNull()

    fun jsonObject(payload: ByteArray, offset: Int, length: Int): JSONObject? =
        runCatching { JSONObject(String(payload, offset, length, Charsets.UTF_8)) }.getOrNull()

    fun jsonBytes(obj: JSONObject): ByteArray = obj.toString().toByteArray(Charsets.UTF_8)
}

class FrameBuffer {
    private var data = ByteArray(256 * 1024)
    private var size = 0

    fun clear() {
        size = 0
    }

    fun append(source: ByteArray, offset: Int, length: Int) {
        if (length <= 0) return
        ensureCapacity(size + length)
        System.arraycopy(source, offset, data, size, length)
        size += length
    }

    fun decodeFrames(): List<Pair<MessageType, ByteArray>> {
        val messages = mutableListOf<Pair<MessageType, ByteArray>>()
        var offset = 0

        while (size - offset >= 9) {
            if (!hasMagicAt(offset)) {
                offset++
                continue
            }

            val type = MessageType.from(data[offset + 4])
            if (type == null) {
                offset++
                continue
            }

            val length = ByteBuffer.wrap(data, offset + 5, 4)
                .order(ByteOrder.BIG_ENDIAN)
                .int
            if (length < 0) {
                offset++
                continue
            }

            val total = 9 + length
            if (size - offset < total) break

            val payload = data.copyOfRange(offset + 9, offset + total)
            messages.add(type to payload)
            offset += total
        }

        if (offset > 0) {
            val remaining = size - offset
            if (remaining > 0) {
                System.arraycopy(data, offset, data, 0, remaining)
            }
            size = remaining
        }

        return messages
    }

    private fun hasMagicAt(offset: Int): Boolean {
        return data[offset] == LinkProtocol.magic[0] &&
            data[offset + 1] == LinkProtocol.magic[1] &&
            data[offset + 2] == LinkProtocol.magic[2] &&
            data[offset + 3] == LinkProtocol.magic[3]
    }

    private fun ensureCapacity(required: Int) {
        if (required <= data.size) return
        var newSize = data.size
        while (newSize < required) {
            newSize *= 2
        }
        data = data.copyOf(newSize)
    }
}

data class PairingPayload(
    val v: Int,
    val host: String,
    val port: Int,
    val token: String
) {
    companion object {
        fun parse(json: String): PairingPayload? = runCatching {
            val obj = JSONObject(json)
            PairingPayload(
                v = obj.getInt("v"),
                host = obj.getString("host"),
                port = obj.getInt("port"),
                token = obj.getString("token")
            )
        }.getOrNull()
    }
}

enum class DeviceSide(val value: String) {
    MAC("mac"),
    ANDROID("android")
}

fun DataOutputStream.writeFrame(type: MessageType, payload: ByteArray = ByteArray(0)) {
    write(LinkProtocol.encode(type, payload))
}

fun DataOutputStream.writeFrameAndFlush(type: MessageType, payload: ByteArray = ByteArray(0)) {
    writeFrame(type, payload)
    flush()
}

File: macos/Sources/SimpleLink/Protocol.swift
import Foundation

enum MessageType: UInt8 {
    case auth = 1
    case authOK = 2
    case authFail = 3
    case clipboard = 4
    case fileBegin = 5
    case fileChunk = 6
    case fileEnd = 7
    case ping = 8
    case pong = 9
}

enum LinkProtocol {
    static let magic = Data("SLNK".utf8)
    static let defaultPort: UInt16 = 9473
    static let chunkSize = 64 * 1024

    static func encode(type: MessageType, payload: Data = Data()) -> Data {
        var frame = Data()
        frame.append(magic)
        frame.append(type.rawValue)
        var length = UInt32(payload.count).bigEndian
        frame.append(Data(bytes: &length, count: 4))
        frame.append(payload)
        return frame
    }

    static func decodeFrames(from buffer: inout Data) -> [(MessageType, Data)] {
        var messages: [(MessageType, Data)] = []

        while buffer.count >= 9 {
            guard buffer.prefix(4) == magic else {
                if let range = buffer.range(of: magic) {
                    buffer.removeSubrange(buffer.startIndex..<range.lowerBound)
                } else {
                    buffer.removeAll()
                }
                continue
            }

            let typeRaw = buffer[4]
            let length = buffer.subdata(in: 5..<9).withUnsafeBytes {
                $0.load(as: UInt32.self).bigEndian
            }
            let total = 9 + Int(length)
            guard buffer.count >= total else { break }

            let payload = buffer.subdata(in: 9..<total)
            if let type = MessageType(rawValue: typeRaw) {
                messages.append((type, payload))
            }
            buffer.removeSubrange(0..<total)
        }

        return messages
    }

    static func jsonData(_ object: [String: Any]) -> Data {
        (try? JSONSerialization.data(withJSONObject: object)) ?? Data()
    }

    static func jsonObject(from data: Data) -> [String: Any]? {
        try? JSONSerialization.jsonObject(with: data) as? [String: Any]
    }
}

struct PairingPayload: Codable {
    let v: Int
    let host: String
    let port: Int
    let token: String
}

enum DeviceSide: String {
    case mac
    case android
}
